In [1]:
import json
import pandas as pd
import numpy as np
import ast
%load_ext autoreload
%autoreload 2
import Funciones
import pyarrow as pa
import pyarrow.parquet as pq
from datetime import datetime
import re

import warnings
warnings.filterwarnings("ignore")

In [2]:
ruta_reviews = r'..\Data\australian_user_reviews.json'

In [3]:
filas_items = []

with open(ruta_reviews, 'r', encoding='utf-8') as f:
    for line in f.readlines():
        filas_items.append(ast.literal_eval(line))

In [4]:
df_reviews = pd.DataFrame(filas_items)
df_reviews

user_id                                           user_url  \
0      76561197970982479  http://steamcommunity.com/profiles/76561197970...   
1                js41637               http://steamcommunity.com/id/js41637   
2              evcentric             http://steamcommunity.com/id/evcentric   
3                  doctr                 http://steamcommunity.com/id/doctr   
4              maplemage             http://steamcommunity.com/id/maplemage   
...                  ...                                                ...   
25794  76561198306599751  http://steamcommunity.com/profiles/76561198306...   
25795           Ghoustik              http://steamcommunity.com/id/Ghoustik   
25796  76561198310819422  http://steamcommunity.com/profiles/76561198310...   
25797  76561198312638244  http://steamcommunity.com/profiles/76561198312...   
25798        LydiaMorley           http://steamcommunity.com/id/LydiaMorley   

                                                 reviews  
0      [{'funny': '', 'posted': 'Posted November 5, 2...  
1      [{'funny': '', 'posted': 'Posted June 24, 2014...  
2      [{'funny': '', 'posted': 'Posted February 3.',...  
3      [{'funny': '', 'posted': 'Posted October 14, 2...  
4      [{'funny': '3 people found this review funny',...  
...                                                  ...  
25794  [{'funny': '', 'posted': 'Posted May 31.', 'la...  
25795  [{'funny': '', 'posted': 'Posted June 17.', 'l...  
25796  [{'funny': '1 person found this review funny',...  
25797  [{'funny': '', 'posted': 'Posted July 21.', 'l...  
25798  [{'funny': '1 person found this review funny',...  

[25799 rows x 3 columns]

Se ve qué tipo de datos hay, y se revisa si hay valores nulos:

In [5]:
Funciones.verificar_tipo_datos(df_reviews)

nombre        tipo_datos  porcentaje_no_nulos  porcentaje_nulos  nulos
0   user_id   [<class 'str'>]                100.0               0.0      0
1  user_url   [<class 'str'>]                100.0               0.0      0
2   reviews  [<class 'list'>]                100.0               0.0      0

Habiendo verificado si hay nulos, se procede a verificar si hay elementos duplicados:

In [6]:

Funciones.duplicados_por_columna(df_reviews, 'user_id')


user_id  \
12888               05041129   
5250                05041129   
3133   111222333444555666888   
3134   111222333444555666888   
4139                   29123   
...                      ...   
2721          xXAussieRockXx   
2680             yolofaceguy   
17916            yolofaceguy   
5855               zeroblade   
13975              zeroblade   

                                                user_url  \
12888              http://steamcommunity.com/id/05041129   
5250               http://steamcommunity.com/id/05041129   
3133   http://steamcommunity.com/id/11122233344455566...   
3134   http://steamcommunity.com/id/11122233344455566...   
4139                  http://steamcommunity.com/id/29123   
...                                                  ...   
2721         http://steamcommunity.com/id/xXAussieRockXx   
2680            http://steamcommunity.com/id/yolofaceguy   
17916           http://steamcommunity.com/id/yolofaceguy   
5855              http://steamcommunity.com/id/zeroblade   
13975             http://steamcommunity.com/id/zeroblade   

                                                 reviews  
12888  [{'funny': '', 'posted': 'Posted May 18, 2015....  
5250   [{'funny': '', 'posted': 'Posted May 18, 2015....  
3133   [{'funny': '', 'posted': 'Posted December 22, ...  
3134   [{'funny': '', 'posted': 'Posted December 22, ...  
4139   [{'funny': '', 'posted': 'Posted March 26.', '...  
...                                                  ...  
2721   [{'funny': '', 'posted': 'Posted July 17, 2015...  
2680   [{'funny': '', 'posted': 'Posted October 31, 2...  
17916  [{'funny': '', 'posted': 'Posted October 31, 2...  
5855   [{'funny': '', 'posted': 'Posted November 30, ...  
13975  [{'funny': '', 'posted': 'Posted November 30, ...  

[623 rows x 3 columns]

Se puede observar que hay 623 filas duplicadas. Se procede a la eliminación de duplicados.

In [7]:
# Se eliminan 
df_reviews = df_reviews.drop_duplicates('user_id', keep=False)



In [8]:
# Se vuelve a ver si sigue habiendo duplicados
Funciones.duplicados_por_columna(df_reviews, 'user_id')

'No hay duplicados'

Se observa la columna 'reviews':

In [9]:
df_reviews['reviews'][0]

[{'funny': '',
  'posted': 'Posted November 5, 2011.',
  'last_edited': '',
  'item_id': '1250',
  'helpful': 'No ratings yet',
  'recommend': True,
  'review': 'Simple yet with great replayability. In my opinion does "zombie" hordes and team work better than left 4 dead plus has a global leveling system. Alot of down to earth "zombie" splattering fun for the whole family. Amazed this sort of FPS is so rare.'},
 {'funny': '',
  'posted': 'Posted July 15, 2011.',
  'last_edited': '',
  'item_id': '22200',
  'helpful': 'No ratings yet',
  'recommend': True,
  'review': "It's unique and worth a playthrough."},
 {'funny': '',
  'posted': 'Posted April 21, 2011.',
  'last_edited': '',
  'item_id': '43110',
  'helpful': 'No ratings yet',
  'recommend': True,
  'review': 'Great atmosphere. The gunplay can be a bit chunky at times but at the end of the day this game is definitely worth it and I hope they do a sequel...so buy the game so I get a sequel!'}]

Este archivo contiene las siguientes columnas:

•user_id: Es el código identificador único del usuario.
•user_url: Es la URL del perfil del usuario.
•reviews: Review de usuario en formato Json
    •funny: Aquí se ve si algún usuario puso algo gracioso.
    •posted: Fecha de posteo de la review.
    •last_edited: Fecha de la última edición.
    •item_id: Es el código identificador único del juego.
    •helpful: Aquí los demás usuarios aclaran si les fue útil la información.
    •recommend: Indica si el usuario recomienda o no recomienda el juego.
    •review: Aquí se ven los comentarios acerca del juego.

Se transforma y trata la columna 'reviews', que se encuentra añidada, se normaliza:

In [10]:
df_reviews_desañidado = pd.json_normalize(df_reviews['reviews'])
df_reviews_desañidado

0  \
0      {'funny': '', 'posted': 'Posted November 5, 20...   
1      {'funny': '', 'posted': 'Posted June 24, 2014....   
2      {'funny': '', 'posted': 'Posted February 3.', ...   
3      {'funny': '', 'posted': 'Posted October 14, 20...   
4      {'funny': '3 people found this review funny', ...   
...                                                  ...   
25171  {'funny': '', 'posted': 'Posted May 31.', 'las...   
25172  {'funny': '', 'posted': 'Posted June 17.', 'la...   
25173  {'funny': '1 person found this review funny', ...   
25174  {'funny': '', 'posted': 'Posted July 21.', 'la...   
25175  {'funny': '1 person found this review funny', ...   

                                                       1  \
0      {'funny': '', 'posted': 'Posted July 15, 2011....   
1      {'funny': '', 'posted': 'Posted September 8, 2...   
2      {'funny': '', 'posted': 'Posted December 4, 20...   
3      {'funny': '', 'posted': 'Posted July 28, 2012....   
4      {'funny': '1 person found this review funny', ...   
...                                                  ...   
25171                                               None   
25172                                               None   
25173                                               None   
25174  {'funny': '', 'posted': 'Posted July 10.', 'la...   
25175  {'funny': '', 'posted': 'Posted July 20.', 'la...   

                                                       2  \
0      {'funny': '', 'posted': 'Posted April 21, 2011...   
1      {'funny': '', 'posted': 'Posted November 29, 2...   
2      {'funny': '', 'posted': 'Posted November 3, 20...   
3      {'funny': '', 'posted': 'Posted June 2, 2012.'...   
4      {'funny': '2 people found this review funny', ...   
...                                                  ...   
25171                                               None   
25172                                               None   
25173                                               None   
25174  {'funny': '', 'posted': 'Posted July 10.', 'la...   
25175  {'funny': '', 'posted': 'Posted July 2.', 'las...   

                                                       3  \
0                                                   None   
1                                                   None   
2      {'funny': '', 'posted': 'Posted October 15, 20...   
3      {'funny': '', 'posted': 'Posted June 29, 2014....   
4      {'funny': '', 'posted': 'Posted July 11, 2013....   
...                                                  ...   
25171                                               None   
25172                                               None   
25173                                               None   
25174  {'funny': '', 'posted': 'Posted July 8.', 'las...   
25175                                               None   

                                                       4  \
0                                                   None   
1                                                   None   
2      {'funny': '', 'posted': 'Posted October 15, 20...   
3      {'funny': '', 'posted': 'Posted November 22, 2...   
4                                                   None   
...                                                  ...   
25171                                               None   
25172                                               None   
25173                                               None   
25174                                               None   
25175                                               None   

                                                       5     6     7     8  \
0                                                   None  None  None  None   
1                                                   None  None  None  None   
2      {'funny': '', 'posted': 'Posted October 15, 20...  None  None  None   
3      {'funny': '', 'posted': 'Posted February 23, 2...  None  None  None   
4                                                   None  No

Se procede a unir lo anterior con las columnas de 'user_id', y 'user_url':

In [11]:
df_reviews_desañidado = pd.concat([df_reviews[['user_id', 'user_url']], df_reviews_desañidado], axis=1)
df_reviews_desañidado

user_id                                           user_url  \
0      76561197970982479  http://steamcommunity.com/profiles/76561197970...   
1                js41637               http://steamcommunity.com/id/js41637   
2              evcentric             http://steamcommunity.com/id/evcentric   
3                  doctr                 http://steamcommunity.com/id/doctr   
4              maplemage             http://steamcommunity.com/id/maplemage   
...                  ...                                                ...   
17819                NaN                                                NaN   
17916                NaN                                                NaN   
18028                NaN                                                NaN   
18234                NaN                                                NaN   
18309                NaN                                                NaN   

                                                       0  \
0      {'funny': '', 'posted': 'Posted November 5, 20...   
1      {'funny': '', 'posted': 'Posted June 24, 2014....   
2      {'funny': '', 'posted': 'Posted February 3.', ...   
3      {'funny': '', 'posted': 'Posted October 14, 20...   
4      {'funny': '3 people found this review funny', ...   
...                                                  ...   
17819  {'funny': '', 'posted': 'Posted April 11.', 'l...   
17916  {'funny': '', 'posted': 'Posted December 29, 2...   
18028  {'funny': '', 'posted': 'Posted June 23, 2014....   
18234  {'funny': '7 people found this review funny', ...   
18309  {'funny': '', 'posted': 'Posted October 31, 20...   

                                                       1  \
0      {'funny': '', 'posted': 'Posted July 15, 2011....   
1      {'funny': '', 'posted': 'Posted September 8, 2...   
2      {'funny': '', 'posted': 'Posted December 4, 20...   
3      {'funny': '', 'posted': 'Posted July 28, 2012....   
4      {'funny': '1 person found this review funny', ...   
...                                                  ...   
17819                                               None   
17916  {'funny': '', 'posted': 'Posted June 25, 2015....   
18028                                               None   
18234  {'funny': '', 'posted': 'Posted June 13, 2014....   
18309                                               None   

                                                       2  \
0      {'funny': '', 'posted': 'Posted April 21, 2011...   
1      {'funny': '', 'posted': 'Posted November 29, 2...   
2      {'funny': '', 'posted': 'Posted November 3, 20...   
3      {'funny': '', 'posted': 'Posted June 2, 2012.'...   
4      {'funny': '2 people found this review funny', ...   
...                                                  ...   
17819                                               None   
17916                                               None   
18028                                               None   
18234  {'funny': '', 'posted': 'Posted May 14, 2014.'...   
18309                                               None   

                                                       3  \
0                                                   None   
1                                                   None   
2      {'funny': '', 'posted': 'Posted October 15, 20...   
3      {'funny': '', 'posted': 'Posted June 29, 2014....   
4      {'funny': '', 'posted': 'Posted July 11, 2013....   
...                                                  ...   
17819                                               None   
17916                                               None   
18028                                               None   
18234  {'funny': '', 'posted': 'Posted June 5, 2014.'...   
18309                                               None   

                                                       4  \
0                                                   None   
1                                                   None   
2   

In [12]:
Funciones.verificar_tipo_datos(df_reviews_desañidado)

nombre                                         tipo_datos  \
0    user_id                   [<class 'str'>, <class 'float'>]   
1   user_url                   [<class 'str'>, <class 'float'>]   
2          0  [<class 'dict'>, <class 'NoneType'>, <class 'f...   
3          1  [<class 'dict'>, <class 'NoneType'>, <class 'f...   
4          2  [<class 'dict'>, <class 'NoneType'>, <class 'f...   
5          3  [<class 'NoneType'>, <class 'dict'>, <class 'f...   
6          4  [<class 'NoneType'>, <class 'dict'>, <class 'f...   
7          5  [<class 'NoneType'>, <class 'dict'>, <class 'f...   
8          6  [<class 'NoneType'>, <class 'dict'>, <class 'f...   
9          7  [<class 'NoneType'>, <class 'dict'>, <class 'f...   
10         8  [<class 'NoneType'>, <class 'dict'>, <class 'f...   
11         9  [<class 'NoneType'>, <class 'dict'>, <class 'f...   

    porcentaje_no_nulos  porcentaje_nulos  nulos  
0                 97.59              2.41    623  
1                 97.59              2.41    623  
2                 97.48              2.52    650  
3                 45.56             54.44  14044  
4                 27.82             72.18  18622  
5                 18.14             81.86  21120  
6                 12.44             87.56  22589  
7                  8.73             91.27  23546  
8                  5.92             94.08  24272  
9                  3.83             96.17  24811  
10                 2.23             97.77  25224  
11                 0.99             99.01  25544

Usamos la función melt para transformar las columnas en filas.  La función melt() toma el DataFrame original y especifica las columnas 'user_id' y 'user_url' como identificadores únicos (id_vars). La columna 'reviews' contiene las reviews correspondientes. 

In [13]:
df_reviews_desañidado = pd.melt(df_reviews_desañidado, id_vars=['user_id', 'user_url'], 
                       value_vars=list(range(9)),
                       value_name='reviews')
df_reviews_desañidado

user_id                                           user_url  \
0       76561197970982479  http://steamcommunity.com/profiles/76561197970...   
1                 js41637               http://steamcommunity.com/id/js41637   
2               evcentric             http://steamcommunity.com/id/evcentric   
3                   doctr                 http://steamcommunity.com/id/doctr   
4               maplemage             http://steamcommunity.com/id/maplemage   
...                   ...                                                ...   
232186                NaN                                                NaN   
232187                NaN                                                NaN   
232188                NaN                                                NaN   
232189                NaN                                                NaN   
232190                NaN                                                NaN   

       variable                                            reviews  
0             0  {'funny': '', 'posted': 'Posted November 5, 20...  
1             0  {'funny': '', 'posted': 'Posted June 24, 2014....  
2             0  {'funny': '', 'posted': 'Posted February 3.', ...  
3             0  {'funny': '', 'posted': 'Posted October 14, 20...  
4             0  {'funny': '3 people found this review funny', ...  
...         ...                                                ...  
232186        8                                               None  
232187        8                                               None  
232188        8                                               None  
232189        8  {'funny': '', 'posted': 'Posted August 16, 201...  
232190        8                                               None  

[232191 rows x 4 columns]

Se eliminan los valores None que posee la columna 'reviews':

In [14]:
df_reviews_desañidado=df_reviews_desañidado.dropna()

Usamos la función json_normalize para expandir los diccionarios en la columna 'reviews' en varias columnas:

In [15]:
df_reviews = pd.json_normalize(df_reviews_desañidado['reviews']).add_prefix('reviews_')
df_reviews = pd.concat([df_reviews_desañidado[['user_id', 'user_url']], df_reviews], axis=1)
df_reviews

user_id                                           user_url  \
0      76561197970982479  http://steamcommunity.com/profiles/76561197970...   
1                js41637               http://steamcommunity.com/id/js41637   
2              evcentric             http://steamcommunity.com/id/evcentric   
3                  doctr                 http://steamcommunity.com/id/doctr   
4              maplemage             http://steamcommunity.com/id/maplemage   
...                  ...                                                ...   
55735                NaN                                                NaN   
55736                NaN                                                NaN   
55740                NaN                                                NaN   
55741                NaN                                                NaN   
55742                NaN                                                NaN   

                          reviews_funny             reviews_posted  \
0                                         Posted November 5, 2011.   
1                                            Posted June 24, 2014.   
2                                               Posted February 3.   
3                                         Posted October 14, 2013.   
4      3 people found this review funny     Posted April 15, 2014.   
...                                 ...                        ...   
55735                                    Posted December 20, 2013.   
55736                                     Posted February 3, 2014.   
55740                                       Posted August 2, 2014.   
55741                                        Posted July 31, 2015.   
55742                                    Posted December 20, 2015.   

                 reviews_last_edited reviews_item_id  \
0                                               1250   
1                                             251610   
2                                             248820   
3                                             250320   
4                                             211420   
...                              ...             ...   
55735  Last edited January 16, 2015.          105600   
55736                                         221910   
55740                                         304930   
55741                                         265630   
55742                                         304050   

                                       reviews_helpful reviews_recommend  \
0                                       No ratings yet              True   
1      15 of 20 people (75%) found this review helpful              True   
2                                       No ratings yet              True   
3       2 of 2 people (100%) found this review helpful              True   
4      35 of 43 people (81%) found this review helpful              True   
...                                                ...               ...   
55735                                   No ratings yet              True   
55736                                   No ratings yet              True   
55740                                   No ratings yet              True   
55741                                   No ratings yet              True   
55742                                   No ratings yet              True   

                                          reviews_review  
0      Simple yet with great replayability. In my opi...  
1      I know what you think when you see this title ...  
2      A suitably punishing roguelike platformer.  Wi...  
3      This game... is so fun. The fight sequences ha...  
4                                                Git gud  
...                                                  ...  
55735                                                  ,  
55736  A first person game which has no guns, no line...  
55740                              Fun game with friends  
55741                                        So Fun

Se reemplazan los valores vacíos de la columna 'reviews_last_edited' como None:

In [16]:
df_reviews.replace('', None, inplace=True)
df_reviews

user_id                                           user_url  \
0      76561197970982479  http://steamcommunity.com/profiles/76561197970...   
1                js41637               http://steamcommunity.com/id/js41637   
2              evcentric             http://steamcommunity.com/id/evcentric   
3                  doctr                 http://steamcommunity.com/id/doctr   
4              maplemage             http://steamcommunity.com/id/maplemage   
...                  ...                                                ...   
55735                NaN                                                NaN   
55736                NaN                                                NaN   
55740                NaN                                                NaN   
55741                NaN                                                NaN   
55742                NaN                                                NaN   

                          reviews_funny             reviews_posted  \
0                                  None   Posted November 5, 2011.   
1                                  None      Posted June 24, 2014.   
2                                  None         Posted February 3.   
3                                  None   Posted October 14, 2013.   
4      3 people found this review funny     Posted April 15, 2014.   
...                                 ...                        ...   
55735                              None  Posted December 20, 2013.   
55736                              None   Posted February 3, 2014.   
55740                              None     Posted August 2, 2014.   
55741                              None      Posted July 31, 2015.   
55742                              None  Posted December 20, 2015.   

                 reviews_last_edited reviews_item_id  \
0                               None            1250   
1                               None          251610   
2                               None          248820   
3                               None          250320   
4                               None          211420   
...                              ...             ...   
55735  Last edited January 16, 2015.          105600   
55736                           None          221910   
55740                           None          304930   
55741                           None          265630   
55742                           None          304050   

                                       reviews_helpful reviews_recommend  \
0                                       No ratings yet              True   
1      15 of 20 people (75%) found this review helpful              True   
2                                       No ratings yet              True   
3       2 of 2 people (100%) found this review helpful              True   
4      35 of 43 people (81%) found this review helpful              True   
...                                                ...               ...   
55735                                   No ratings yet              True   
55736                                   No ratings yet              True   
55740                                   No ratings yet              True   
55741                                   No ratings yet              True   
55742                                   No ratings yet              True   

                                          reviews_review  
0      Simple yet with great replayability. In my opi...  
1      I know what you think when you see this title ...  
2      A suitably punishing roguelike platformer.  Wi...  
3      This game... is so fun. The fight sequences ha...  
4                                                Git gud  
...                                                  ...  
55735                                                  ,  
55736  A first person game which has no guns, no line...  
55740                              Fun game with friends  
55741                                        So Fun

In [17]:
Funciones.verificar_tipo_datos(df_reviews)

nombre                                         tipo_datos  \
0              user_id                   [<class 'str'>, <class 'float'>]   
1             user_url                   [<class 'str'>, <class 'float'>]   
2        reviews_funny  [<class 'NoneType'>, <class 'str'>, <class 'fl...   
3       reviews_posted                   [<class 'str'>, <class 'float'>]   
4  reviews_last_edited  [<class 'NoneType'>, <class 'str'>, <class 'fl...   
5      reviews_item_id                   [<class 'str'>, <class 'float'>]   
6      reviews_helpful                   [<class 'str'>, <class 'float'>]   
7    reviews_recommend                  [<class 'bool'>, <class 'float'>]   
8       reviews_review  [<class 'str'>, <class 'NoneType'>, <class 'fl...   

   porcentaje_no_nulos  porcentaje_nulos  nulos  
0                75.19             24.81  18395  
1                75.19             24.81  18395  
2                10.32             89.68  66485  
3                75.19             24.81  18395  
4                 7.70             92.30  68431  
5                75.19             24.81  18395  
6                75.19             24.81  18395  
7                75.19             24.81  18395  
8                75.15             24.85  18423

Se eliminan las columnas 'reviews_funny' y 'reviews_last_edited' por que al tener muchos valores faltantes, nos resulta engorroso y poco útil conservarlas:


In [18]:
df_reviews = df_reviews.drop(columns=['reviews_funny', 'reviews_last_edited'])
df_reviews.columns

Index(['user_id', 'user_url', 'reviews_posted', 'reviews_item_id',
       'reviews_helpful', 'reviews_recommend', 'reviews_review'],
      dtype='object')

Se transforma la columna 'reviews_posted', y se extrae el año:

In [19]:

df_reviews['reviews_posted'].fillna('sin fecha', inplace=True)

In [20]:
df_reviews['reviews_posted'] = df_reviews['reviews_posted'].apply(Funciones.extraer_anio)

In [21]:
df_reviews.head()

user_id                                           user_url  \
0  76561197970982479  http://steamcommunity.com/profiles/76561197970...   
1            js41637               http://steamcommunity.com/id/js41637   
2          evcentric             http://steamcommunity.com/id/evcentric   
3              doctr                 http://steamcommunity.com/id/doctr   
4          maplemage             http://steamcommunity.com/id/maplemage   

  reviews_posted reviews_item_id  \
0           2011            1250   
1           2014          251610   
2              3          248820   
3           2013          250320   
4           2014          211420   

                                   reviews_helpful reviews_recommend  \
0                                   No ratings yet              True   
1  15 of 20 people (75%) found this review helpful              True   
2                                   No ratings yet              True   
3   2 of 2 people (100%) found this review helpful              True   
4  35 of 43 people (81%) found this review helpful              True   

                                      reviews_review  
0  Simple yet with great replayability. In my opi...  
1  I know what you think when you see this title ...  
2  A suitably punishing roguelike platformer.  Wi...  
3  This game... is so fun. The fight sequences ha...  
4                                            Git gud

Se eliminan nulos:

In [22]:
#Se eliminan los nulos
df_reviews = df_reviews.dropna(subset=['reviews_review'])

In [23]:
df_reviews = df_reviews.dropna(subset=['user_id'])

In [24]:
df_reviews = df_reviews.dropna(subset=['user_url'])

Se guarda el dataframe final:

In [25]:
archivo_csv = r'..\Data\user_reviews.csv'
df_reviews.to_csv(archivo_csv, index=False, encoding='utf-8')

In [26]:
table = pa.Table.from_pandas(df_reviews)

In [27]:
nombre_archivo_parquet = 'users_reviews.parquet'

In [28]:
pq.write_table(table, nombre_archivo_parquet)